src/components/CodeSnippet/CodeSnippet.js

```diff
@@ -1,43 +1,159 @@
+     */
+    light: PropTypes.bool,
+  };
+
+  static defaultProps = {
+    type: 'single',
+    showMoreText: 'Show more',
+    showLessText: 'Show less',
+  };
+
+  state = {
+    showBtn: false,
+    expandedCode: false,
+  };
+
+  componentDidMount = () => {
+    if (this.codeContent) {
+      if (this.codeContent.getBoundingClientRect().height > 255) {
+        this.setState({ showBtn: true });
+      }
+    }
+  };
+
+  expandCode = () => {
+    this.setState({ expandedCode: !this.state.expandedCode });
+  };
+
+  render() {
+    const {
+      className,
+      type,
+      children,
+      feedback,
+      onClick,
+      copyLabel,
+      light,
+      showMoreText,
+      showLessText,
+      ...other
+    } = this.props;
+
+    const codeSnippetClasses = classNames(className, {
+      'bx--snippet': true,
+      'bx--snippet--single': type === 'single' || type === 'terminal',
+      'bx--snippet--multi': type === 'multi' || type === 'code',
+      'bx--snippet--inline': type === 'inline',
+      'bx--snippet--expand': this.state.expandedCode,
+      'bx--snippet--light': light,
+    });
+
+    const more = showMoreText;
+    const less = showLessText;
+
+    const expandCodeBtnText = this.state.expandedCode ? less : more; // need the show/more less text props to show here
+
+    const moreLessBtn = (
+      <button
+        className="bx--btn bx--btn--ghost bx--btn--sm bx--snippet-btn--expand"
+        type="button"
+        onClick={this.expandCode}>
+        <span className="bx--snippet-btn--text">{expandCodeBtnText}</span>
+        <Icon
+          alt={expandCodeBtnText}
+          name="chevron--down"
+          description={expandCodeBtnText}
+          className="bx--icon-chevron--down"
+        />
+      </button>
+    );
+
+    const code = (
       <div role="textbox" tabIndex={0} className="bx--snippet-container">
         <code>
-          <pre ref={wrappedContentRef}>{children}</pre>
+          <pre
+            ref={codeContent => {
+              this.codeContent = codeContent;
+            }}>
+            {children}
+          </pre>
         </code>
       </div>
-      <CopyButton onClick={onClick} feedback={feedback} />
-    </div>
-  );
-};
-
-CodeSnippet.propTypes = {
-  type: PropTypes.string,
-  className: PropTypes.string,
-  children: PropTypes.string,
-  feedback: PropTypes.string,
-  onClick: PropTypes.func,
-  wrappedContentRef: PropTypes.func,
-};
-
-CodeSnippet.defaultProps = {
-  type: 'terminal',
-};
-
-export default CodeSnippet;
+    );
+
+    const copy = <CopyButton onClick={onClick} feedback={feedback} />;
+
+    if (type === 'inline' || type === 'terminal') {
+      return (
+        <Copy
+          className={codeSnippetClasses}
+          aria-label={copyLabel}
+          feedback={feedback}>
+          <code>{children}</code>
+        </Copy>
+      );
+    }
+
+    if (type === 'single') {
+      return (
+        <div className={codeSnippetClasses} {...other}>
```

Comment:
Should we re-order this to:

```diff
-        <div className={codeSnippetClasses} {...other}>
+        <div {...other} className={codeSnippetClasses}>
```

Just to make sure `className` doesn't get overridden?